File: Lista_1/Lista_2.scala

zad 2
def fib (p:Int):Int =
  p match
  {case  0 =>0
   case  1 =>1
   case  _ =>fib(p-1)+fib(p-2)
     
  }
fib(42)
def fibpom (a:Int, b:Int, n:Int):Int=
  n match
  {
  case 0 =>a
  case _ => fibpom(b, a+b, n-1)
  }
def fibn (n:Int) = fibpom(0, 1, n);
fibn (42);
zad 3

def root3(a:Double, e:Double):Double = 
  a>1 match
    {
    case (true) => root3Pom(a, a/3, e)
    case (false) =>root3Pom(a, a, e)
    }
 
 	
	def root3Pom(a:Double, x:Double, e:Double):Double=
	  (absF(Math.pow(x,3)-a)<=e*absF(a)) match
	  {
	  case (true) =>x
	  case (false) => root3Pom (a, (x+(a/(x*x)-x)/3), e)
	  }
	
	def absF (x:Double):Double=
	 x>=0 match
	{
	   case true => x
	   case false => -x
	} 
	
root3(1000, Math.pow(10, -15))


zad 4
val a =List (-2, -1, 0, 1, 2)
val List (_, _, x, _, _) = a

val b = List((1,2),(0,1))
val List((_,_),(x,_)) = b

zad 5

def initSegment[A,B](xs:List[A], xm:List[B]):Boolean =
  (xs, xm) match
  {
  case (Nil, _) =>true
  case (_, Nil) => false
  case (h1::t1,h2::t2) if h1==h2 =>initSegment(t1, t2)
  case _ => false
  }
	initSegment(List(3,4,5,6,7), List(3,4,5,6,7,8,9))
	
zad 6

def replaceNth[A](ls:List[A], l:Int, z:A):List[A]=
  (ls,l) match
  {
  case (Nil, _) =>Nil
  case (hd::tl,0) =>z::tl
  case (hd::tl, _) => hd::replaceNth(tl, l-1, z)
  }
	
		replaceNth(5::8::9::1::3::4::5::Nil,2, 100);;

File: Lista_1/Lista_2.ml

zad 2
let rec fib p= 
	match p with
	| 0->0
	| 1->1
	| _ -> fib(p-1)+fib(p-2) ;;

fib(42);;

let rec fibpom (a, b, n) = 
	match n with
	| 0 -> a
	| _ ->fibpom (b, (a + b), (n - 1));;
	

let fibn n = fibpom (0, 1, n);;
fibn 42;;


zad 3
let rec root3 (a, e)=
	match (a>1.) with
	| true -> root3pom(a, a/.3., e)
	| false ->root3pom(a, a, e);;

 

let rec root3pom (a, x, e) = 
	match (absF((x**3.)-.a)<=e*.absF(a)) with
	| true -> x
	| false ->root3pom (a, x+.((a/.(x**2.))-.x)/.3., e);;

let absF (x)=
	match x>=0. with
	| true -> x
	| false -> -.x ;;

	root3(1000., 10.**(-1.));;


zad 4
let a = [-2; -1; 0; 1; 2];;
let [_; _; x; _; _] = a;;

let b = [(1,2);(0,1)];;
let [(_,_);(x,_)] = b;;

zad 5
let rec initSegment (xs, xm) = 
	match (xs, xm) with 
	| [],_ ->true
	| _,[]->false
	| hd1::tl1,hd2::tl2 when(hd1=hd2) ->initSegment(tl1, tl2)
	| _ -> false;;

	initSegment(3::2::7::[], []);;
zad 6

let rec replaceNth (ls, l, z) =
	match (ls, l) with
	| [], _ ->[]
	| hd::tl,0 ->z::tl
	| hd::tl, _->hd::replaceNthhh(tl, l-1, z);; 
